plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,251 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2020 The Fluent Bit Authors
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_output.h>
+#include <fluent-bit/flb_pack.h>
+
+#include "pgsql.h"
+
+void pgsql_conf_destroy(struct flb_pgsql_config *ctx)
+{
+    PGresult *res = NULL;
+
+    if(PQstatus(ctx->conn) == CONNECTION_OK) {
+        while(PQconsumeInput(ctx->conn) == 0) {
+            res = PQgetResult(ctx->conn);
+            if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+                flb_warn("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+            }
+            PQclear(res);
+        }
+    }
+
+    flb_free(ctx->db_hostname);
+    flb_sds_destroy(ctx->db_table);
+    flb_sds_destroy(ctx->timestamp_key);
+
```

Comment:
For all variables in the context, it's probably good to check if they are null or not before freeing them. In the init functions you can return if `db_table` could not be allocated (so it will be null), if you then free it here that will result in a segfault.

Reply:
In case of a null flb_free() will do nothing
In case of null flb_sds_destory() will do nothing

So I think it's already covered.

Reply:
> In case of a null flb_free() will do nothing

I believe this is the case

> In case of null flb_sds_destory() will do nothing

In my experience this is not the case. Look at the code for SDS strings. When they are allocated there's a bit of memory ahead of the actual `char *` that contains metadata. When the string is destroyed, you pass it the `char *` and it does pointer math to find the actual start of the memory block. I've written similar code in the past that can pass it NULL and it led to segfaults.

Reply:
See: https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L348

Reply:
Actually never mind, I realized it checks for NULL here: https://github.com/fluent/fluent-bit/blob/master/src/flb_sds.c#L342

@sxd You are correct.

I think the bugs I've faced in the past were from passing it an invalid (but non-null address), since you use calloc to create the context it works.

Reply:
ho! ok got it make lot of sense that! still trying to learn more about the flb_sds interface, I'll add the checks!

Reply:
done! :D